casanovo/casanovo.py

```diff
@@ -382,7 +395,76 @@ def setup_model(
     return config, model
 
 
-def _get_model_weights() -> str:
+def _get_weights_from_url(
+    file_url: Optional[str],
+    cache_dir: Path,
+) -> str:
+    """
+    Attempt to download weight file from URL if weights are not already
+    cached. Otherwise use cased weights. Downloaded weight files will be
+    cached.
+
+    Parameters
+    ----------
+    file_url : str
+        url pointing to model weights file
+    cache_dir : Path
+        model weights cache directory path
+
+    Returns
+    -------
+    str
+        path to cached weights file
+    """
+    os.makedirs(cache_dir, exist_ok=True)
+    url_hash = shake_256(file_url.encode("utf-8")).hexdigest(10)
+    cache_file_name = url_hash + ".ckpt"
```

Comment:
**todo:** I'm not a huge fan of this, because this removes the human-interpretable weights file name that will subsequently be used. And for example the information that will be logged, which will make it much more challenging for us to diagnose problems related to specific models.

Reply:
When I've worked with hashed objects previously normally for human readable documents like reports, logs, or what have you the object is just referred to the first five characters of the hash. Could something like this work here? Another alternative I've thought of is to allow the user to set a human readable alias for the downloaded file, and the file name would become <url_hash>:<alias>.ckpt.

Reply:
My main concern is not the length of the hash or what it looks like, but rather that any hashing removes the information conveyed by the filename. Given that we have different model weights (tryptic, non-tryptic, trained on different subsets, ...) that are distinguished by their filenames, this is quite important information to be include in the log for debugging purposes. So using the first five characters is not a solution.

Allowing the user to set an alias would address this to some extent, but it's not ideal either. Users don't know what they're doing, and they'll just use random names that we won't know what they mean either. So ideally the original filename should be conserved.

Can this be addressed by keeping the filenames as-is, but saving the hash in a separate file in our cache directory? Then this will get consulted to check whether the file needs to be downloaded again.

Reply:
True. How about something like this:

- All of the following are files and subdirectories in the Casanovo cache directory.
- An index file that contains a mapping from hashes to file names.
- The weights files are then stored in a subdirectory under their hash. I.e. `[cache_dir]/[hash]/[original_filname.ckpt]`

This conserves the original file names and avoids filename collisions.